hp_list/main.c

```diff
@@ -14,6 +14,124 @@
+ * and __list_find.
+ * "store" is the number of atomic_store operation in list_delete, list_insert
+ * and __list_find.
+ */
+struct runtime_stat {
+    atomic_uint_fast64_t rtry, cons, trav, fail;
+    atomic_uint_fast64_t del, ins;
+    atomic_uint_fast64_t load, store;
+};
+static struct runtime_stat stats = {0};
+
+enum {
+    TRACE_nothing = 0,
+    TRACE_rtry,
+    TRACE_cons,
+    TRACE_trav,
+    TRACE_del,
+    TRACE_ins,
+    TRACE_inserts,
+    TRACE_deletes
+};
+
+#define CAS(obj, expected, desired)                                          \
+    ({                                                                       \
+        bool __ret = atomic_compare_exchange_strong(obj, expected, desired); \
+        if (!__ret)                                                          \
+            atomic_fetch_add(&stats.fail, 1);                                \
+        __ret;                                                               \
+    })
+#define ATOMIC_LOAD(obj)                  \
+    ({                                    \
+        atomic_fetch_add(&stats.load, 1); \
+        atomic_load(obj);                 \
+    })
+#define ATOMIC_STORE_EXPLICIT(obj, desired, order)  \
+    do {                                            \
+        atomic_fetch_add(&stats.store, 1);          \
+        atomic_store_explicit(obj, desired, order); \
+    } while (0)
+#define TRACE(ops)                           \
+    ({                                       \
+        if (TRACE_##ops)                     \
+            atomic_fetch_add(&stats.ops, 1); \
+    })
+
+static void do_analysis(void)
+{
+    printf(
+        "\"rtry\"    is the number of retries in the __list_find function.\n");
+    printf(
+        "\"cons\"    is the number of wait-free contains in the __list_find "
+        "function that curr pointer pointed.\n");
+    printf(
+        "\"trav\"    is the number of list element traversal in the "
+        "__list_find function.\n");
+    printf("\"fail\"    is the number of CAS() failures.\n");
+    printf(
+        "\"del\"     is the number of list_delete operation failed and "
+        "restart again.\n");
+    printf(
+        "\"ins\"     is the number of list_insert operation failed and "
+        "restart again.\n");
+    printf("\"deletes\" is the number of linked list elements deleted.\n");
+    printf("\"inserts\" is the number of linked list elements created.\n");
+    printf(
+        "\"load\"    is the number of atomic_load operation in list_delete, "
+        "list_insert and __list_find.\n");
+    printf(
+        "\"store\"   is the number of atomic_store operation in list_delete, "
+        "list_insert and __list_find.\n");
+    printf("\n%10s %10s %10s %10s %10s %10s %10s %10s %10s %10s\n", "rtry",
+           "cons", "trav", "fail", "del", "ins", "load", "store", "deletes",
+           "inserts");
+    for (int i = 0; i < 109; i++)
```

Comment:
Is it necessary to use atomic operations here? There should be only single thread running.